dataline-workers/src/main/java/io/dataline/workers/singer/SingerTapFactory.java

```diff
@@ -0,0 +1,151 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SingerTapFactory implements TapFactory<SingerMessage> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SingerTapFactory.class);
+
+  private static final String CONFIG_JSON_FILENAME = "tap_config.json";
+  private static final String CATALOG_JSON_FILENAME = "catalog.json";
+
+  private static final String STATE_JSON_FILENAME = "input_state.json";
+
+  private final String dockerImageName;
+
+  private Process tapProcess = null;
+  private InputStream stdout = null;
+
+  public SingerTapFactory(String dockerImageName) {
+    this.dockerImageName = dockerImageName;
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  @Override
+  public Stream<SingerMessage> create(StandardTapConfig input, Path workspaceRoot)
+      throws InvalidCredentialsException {
+    OutputAndStatus<SingerCatalog> discoveryOutput = runDiscovery(input, workspaceRoot);
+
+    final ObjectMapper objectMapper = new ObjectMapper();
+    final String configDotJson;
+    final String catalogDotJson;
+    final String stateDotJson;
+
+    try {
+      configDotJson =
+          objectMapper.writeValueAsString(
+              input.getSourceConnectionImplementation().getConfiguration());
+      SingerCatalog selectedCatalog =
+          SingerCatalogConverters.applySchemaToDiscoveredCatalog(
+              discoveryOutput.getOutput().get(), input.getStandardSync().getSchema());
+      catalogDotJson = objectMapper.writeValueAsString(selectedCatalog);
+      stateDotJson = objectMapper.writeValueAsString(input.getState());
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+
+    // write config.json to disk
+    Path configPath =
+        WorkerUtils.writeFileToWorkspace(workspaceRoot, CONFIG_JSON_FILENAME, configDotJson);
+    Path catalogPath =
+        WorkerUtils.writeFileToWorkspace(workspaceRoot, CATALOG_JSON_FILENAME, catalogDotJson);
+    Path statePath =
+        WorkerUtils.writeFileToWorkspace(workspaceRoot, STATE_JSON_FILENAME, stateDotJson);
+
+    try {
+
+      String[] tapCmd =
+          DockerUtils.getDockerCommand(
+              workspaceRoot,
+              dockerImageName,
+              "--config",
+              configPath.toString(),
+              // TODO support both --properties and --catalog depending on integration
+              "--properties",
+              catalogPath.toString(),
+              "--state",
+              statePath.toString());
+
+      LOGGER.info("running command: {}", Arrays.toString(tapCmd));
+
+      tapProcess =
+          new ProcessBuilder()
+              .command(tapCmd)
+              .redirectError(workspaceRoot.resolve(DefaultSyncWorker.TAP_ERR_LOG).toFile())
+              .start();
+
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    stdout = tapProcess.getInputStream();
+    return Streams.stream(new SingerJsonIterator(stdout)).onClose(getCloseFunction());
```

Comment:
@jrhizor - would be curious to hear what you think about this. i'm a little wary because 1. people don't usually call close on streams in java 2. it doesn't codify close in the interface so it's a little harder to enforce the close step in the lifecycle. don't hate it. don't love it. what do you think?

Reply:
Seems fine to me. The fact that the close can be called separately in the cancellation project is the part that feels clunky, so I'm not sure changing this makes the overall usage "nicer".